pkg/controller/replica.go

```diff
@@ -0,0 +1,334 @@
+	}
+
+	if azVolumeAttachment.Spec.RequestedRole == v1alpha1.PrimaryRole {
+		// Deletion Event
+		if requestedToBeDeleted(azVolumeAttachment) {
+			// If primary attachment is marked for deletion, queue garbage collection for replica attachments
+			r.triggerGarbageCollection(azVolumeAttachment.Spec.UnderlyingVolume)
+		} else {
+			// If not, cancel scheduled garbage collection if there is one enqueued
+			r.cancelGarbageCollection(azVolumeAttachment.Spec.UnderlyingVolume)
+
+			// If promotion event, create a replacement replica
+			if isAttached(azVolumeAttachment) && azVolumeAttachment.Status.Detail.Role != azVolumeAttachment.Spec.RequestedRole {
+				if err := r.manageReplicas(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err != nil {
+					return reconcile.Result{Requeue: true}, err
+				}
+			}
+		}
+	} else {
+		// create a replacement replica if replica attachment failed or promoted
+		if requestedToBeDeleted(azVolumeAttachment) {
+			if azVolumeAttachment.Annotations == nil || !metav1.HasAnnotation(azVolumeAttachment.ObjectMeta, azureutils.CleanUpAnnotation) {
+				go func() {
+					conditionFunc := func() (bool, error) {
+						var tmp v1alpha1.AzVolumeAttachment
+						err = r.client.Get(ctx, request.NamespacedName, &tmp)
+						return errors.IsNotFound(err), nil
+					}
+					_ = wait.PollImmediateInfinite(azureutils.DefaultAttachTimeout/2, conditionFunc)
+
+					for i := 0; i < maxRetry; i++ {
+						if err := r.manageReplicas(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err == nil {
+							break
+						}
+					}
+				}()
+			}
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileReplica) triggerGarbageCollection(volumeName string) {
+	emptyCtx := context.TODO()
+	deletionCtx, cancelFunc := context.WithCancel(emptyCtx)
+	r.cleanUpMap.add(volumeName, cancelFunc)
+	klog.Infof("garbage collection of AzVolumeAttachments for AzVolume (%s) scheduled in %s.", volumeName, DefaultTimeUntilDeletion.String())
+
+	go func(ctx context.Context) {
+		// Sleep
+		time.Sleep(DefaultTimeUntilDeletion)
+		select {
+		case <-ctx.Done():
+			klog.Infof("garbage collection for AzVolume (%s) cancelled", volumeName)
+			return
+		default:
+			_, _ = cleanUpAzVolumeAttachmentByVolume(ctx, r.client, r.azVolumeClient, r.namespace, volumeName, all)
+		}
+	}(deletionCtx)
+}
+
+func (r *ReconcileReplica) cancelGarbageCollection(volumeName string) {
+	v, ok := r.cleanUpMap.get(volumeName)
+	if ok {
+		r.cleanUpMap.lock()
+		if _, ok = r.cleanUpMap.getWithoutLock(volumeName); ok {
+			cancelFunc := v.(context.CancelFunc)
+			cancelFunc()
+			delete(r.cleanUpMap.mapping, volumeName)
+		}
+		r.cleanUpMap.unlock()
+	}
+}
+
+func (r *ReconcileReplica) manageReplicas(ctx context.Context, volumeName string) error {
+	// this is to prevent multiple sync volume operation to be performed on a single volume concurrently as it can create or delete more attachments than necessary
+	v, ok := r.mutexMap.get(volumeName)
```

Comment:
AFAICT, entries are added by the `AzVolume` controller but only consumed here. If the `AzVolume` controller misses an instance for some reason (not sure how likely a scenario this is), then there will be no synchronization for that `AzVolume`. Since the mutex appears to be simply for serializing replica management here, the split in responsibilities here makes it difficult to understand the usage and could make debugging more difficult. Have you considered simply adding entries here opportunitistically instead? That would make usage of the mutex map less error prone.

Reply:
As you pointed out, currently, the `AzVolume` controller is responsible for updating `mutexMap`, which `Replica` controller consumes to serialize replica management.

As far as I can think of, there shouldn't be any occurrence where entry for volume in `mutexMap` is empty in this scenario. `AzVolumeAttachment` CRIs should not be created without a successful underlying volume creation. So in theory, given that an entry to `mutexMap` is always added before `triggerCreate` is called in `AzVolume` controller, the corresponding entry in `mutexMap` should always exist for the volume when `manageReplica` is being called on it.

But we can certainly add to the map opportunistically in `manageReplica` function instead. I stayed away from doing so as it would require mutex double locking and wasn't sure how much impact it would have in lock efficiency (although I can only think that it would be minimal).

Reply:
I don't believe double-locking would be a significant issue here.